src/scancode/cli.py

```diff
@@ -217,724 +225,891 @@ def print_version(ctx, param, value):
-    scanners = OrderedDict(zip(scanners.keys(), scanner_functions))
+    # FIXME: this path computation is super inefficient
+    # tuples of  (absolute location, resource id)
+    # TODO: should we alk topdown or not???
 
-    if processes:
-        interrupter = interruptible
-    else:
-        # fake, non inteerrupting used for debugging when processes=0
-        interrupter = fake_interruptible
-
-    if any(scanner_functions):
-        # Skip other scans if already cached
-        # FIXME: ENSURE we only do this for files not directories
-        if not resource.is_cached:
-            # run the scan as an interruptiple task
-            scans_runner = partial(scan_one, resource.abs_path, scanners, diag)
-            success, scan_result = interrupter(scans_runner, timeout=timeout)
-            if not success:
-                # Use scan errors as the scan result for that file on failure this is
-                # a top-level error not attachedd to a specific scanner, hence the
-                # "scan" key is used for these errors
-                scan_result = {'scan_errors': [scan_result]}
-
-            scans_cache.put_scan(resource.rel_path, resource.get_info(), scan_result)
-
-            # do not report success if some other errors happened
-            if scan_result.get('scan_errors'):
+    resources = ((r.get_path(absolute=True), r.rid) for r in codebase.walk())
+
+    runner = partial(scan_resource, scanners=scanners, timeout=timeout)
+
+    has_info_scanner = with_info
+    lscan = len(scanners)
+    has_other_scanners = lscan > 1 if has_info_scanner else lscan
+    if TRACE:
+        logger_debug('scan_codebase: scanners:', '\n'.join(repr(s) for s in scanners))
+        logger_debug('scan_codebase: has_other_scanners:', bool(has_other_scanners))
+
+    get_resource = codebase.get_resource
+
+    success = True
+    pool = None
+    scans = None
+    try:
+        if processes:
+            # maxtasksperchild helps with recycling processes in case of leaks
+            pool = get_pool(processes=processes, maxtasksperchild=1000)
+            # Using chunksize is documented as much more efficient in the Python doc.
+            # Yet "1" still provides a better and more progressive feedback.
+            # With imap_unordered, results are returned as soon as ready and out of order.
+            scans = pool.imap_unordered(runner, resources, chunksize=1)
+            pool.close()
+        else:
+            # no multiprocessing with processes=0
+            scans = imap(runner, resources)
+
+        if progress_manager:
+            scans = progress_manager(scans)
+            # hack to avoid using a context manager
+            if hasattr(scans, '__enter__'):
+                scans.__enter__()
+
+        while True:
+            try:
+                if with_timing:
+                    location, rid, scan_errors, scan_result, scan_time, scan_result, scan_timings = scans.next()
```

Comment:
`scan_result` occurs twice. The first one needs to be removed.

Reply:
Thank you!
Fixed and I added tests that were missing for the `--timing` option: it was not working at all before that :(